pkg/cloud/aws/aws.go

```diff
@@ -81,532 +73,59 @@ func NewAWSProvider(
 		nattPort = helpers.SubmarinerNatTPort
 	}
 
-	awsClient, err := client.NewClient(kubeClient, clusterName, credentialsSecretName, region)
+	deployer := &machineSetDeployer{workClient: workClient, clusterName: clusterName}
+	ec2Client, err := newEC2Client(kubeClient, clusterName, credentialsSecretName, region)
 	if err != nil {
 		return nil, err
 	}
+	awsCloud := prepareaws.NewCloud(deployer, ec2Client, infraId, region, instanceType)
+
+	input := prepareapi.PrepareForSubmarinerInput{
+		InternalPorts: []prepareapi.PortSpec{
+			{Port: helpers.SubmarinerMetricsPort, Protocol: "tcp"},
+		},
+		PublicPorts: []prepareapi.PortSpec{
+			{Port: uint16(nattPort), Protocol: "udp"},
+			{Port: uint16(ikePort), Protocol: "udp"},
+		},
+		Gateways: 1,
+	}
 
 	return &awsProvider{
-		workClinet:    workClient,
-		awsClinet:     awsClient,
+		awsCloud:      awsCloud,
+		input:         input,
+		reporter:      reporter.NewCloudPrepareReporter(),
 		eventRecorder: eventRecorder,
-		region:        region,
-		infraId:       infraId,
-		ikePort:       int64(ikePort),
-		nattPort:      int64(nattPort),
-		clusterName:   clusterName,
 	}, nil
 }
 
 // PrepareSubmarinerClusterEnv prepares submariner cluster environment on AWS
-// The below tasks will be executed
-// 1. open submariner route port (4800/UDP) between all master and worker nodes
-// 2. open submariner metrics port (8080/TCP) between all master and worker nodes
-// 3. open IPsec ports (by default, 4500/UDP and 500/UDP) for submariner gateway instances
-// 4. find one pulic subnet and tag it with label AWS internal elb label for automatic
-//    subnet discovery by aws load balancers or ingress controllers
-// 5. apply a manifest work to create a MachineSet on managed cluster to create a new AWS
-//    instance for submariner gateway
 func (a *awsProvider) PrepareSubmarinerClusterEnv() error {
```

Comment:
finally, we will remove our interface (after the gcp part is merged to cloud-prepare lib),  I think we only need to have a factory, right?